File: python/computer.py
from bluetooth import BluetoothDevice
from input_devices import InputDevice
from memory import Memory
from output_devices import OutputDevice
from processors import Processor
from wifi import WiFiDevice
from brands import Brand

class Computer:
    """
    The responsibility of this Computer class is to control other Peripherals.
    And control is the only reason why Computer would change.
    """

    def __init__(self, input_device: InputDevice, processor: Processor, memory: Memory, output_device: OutputDevice,brand:Brand):
        self._input_device = input_device
        self._processor = processor
        self._memory = memory
        self._output_device = output_device
        self._brand = brand

    def input(self):
        self._input_device.input()

    def set_input_device(self, input_device: InputDevice):
        self._input_device = input_device

    def get_input_device(self):
        return self._input_device

    def process(self, data: any, instruction_id: int) -> bool:
        return self._processor.process(data, instruction_id)

    def set_memory(self, memory: Memory):
        self._memory = memory

    def get_memory(self):
        return self._memory

    def store(self, data: any, key: str):
        return self._memory.store(data, key)

    def retrieve(self, key: str):
        return self._memory.retrieve(key)

    def set_processor(self, processor: Processor):
        self._processor = processor

    def get_processor(self):
        return self._processor

    def output(self, data: any):
        self._output_device.output(data)

    def set_output_device(self, output_device: OutputDevice):
        self._output_device = output_device

    def get_output_device(self):
        return self._output_device
    
    def boot(self):
        self._brand.boot()
    


class Desktop(Computer):
    pass
class Laptop(Computer):
    pass
 
class Walltop(Computer):
    pass

# Usage

File: typescript/computer.ts
import { BluetoothDevice } from "./bluetooth";
import InputDevice from "./inputDevices"
import Memory from "./memory";
import OutputDevice from "./outputDevices";
import Processor from "./processors";
import { WiFiDevice } from "./WiFi";
import Brand, {HP, Dell, IBM, Lenovo } from "./brands";

/**
    The Responsibility of this Computer class is to control other Peripherals.
    And control is the only rason why Coputer would change.
**/
export default class Computer{

    // Hide computer Fields/Properties from the public
        private inputDevice:InputDevice;
        private processor:Processor;
        private memory:Memory;
        private outputDevice:OutputDevice;
        private brand:Brand;


    // Computer's default settings/configurations using the constructor function.
        constructor(
            inputDevice:InputDevice,
            processor:Processor,
            memory:Memory,
            outputDevice:OutputDevice,
            brand:Brand
        ){
            this.inputDevice = inputDevice;
            this.processor = processor;
            this.memory = memory;
            this.outputDevice = outputDevice;
            this.brand = brand;
        }

    // Control Input Devices
        input(){
        this.inputDevice.input();
        }

        setInputDevice(inputDevice:InputDevice){
            this.inputDevice = inputDevice;
        }

        getInputDevice(){
            return this.inputDevice
        }
        // Creating method boot 

        boot(){
            this.brand.boot();
        }

    

    // Control Processing Devices
        process(data:any, instructionId:number): boolean{
            return this.processor.process(data, instructionId);
        }
    
        setMemory(memory:Memory){
            this.memory = memory;
        }

        getMemory(){
            return this.memory
        }


     // Control Storage Devices
        store(data: any, key: string){
            return this.memory.store(data, key);
        }

        retrieve(key:string):any{
           return this.memory.retrieve(key);
        }

        setProcessor(processor:Processor){
            this.processor = processor;
        }

        getProcessor(){
            return this.processor
        }



    // Control Output Devices
        output(data: any){
            this.outputDevice.output(data);
        }

        setOutputDevice(outputDevice:OutputDevice){
            this.outputDevice = outputDevice;
        }

        getOutputDevice(){
            return this.outputDevice;
        }
}

// Laptop is a Computer that supports WiFi and Bluetooth technologies.
export class Laptop extends Computer implements WiFiDevice, BluetoothDevice{
    connectToWIFI(){
        console.log('Connecting to wifi.');
        return true;
    }

    connectToBluetooth(){
        console.log('Connecting to Bluetooth.');
        return true;
    }

    // added functionality
    fold(){
        console.log("Folding...");       
    }
}

//1. Desktop

class Desktop extends Computer{
    // we define a constructor  through which we shall pass the desktop benad 

}

//3. Walltop 
class Walltop extends Computer{
    // we define a constructor  through which we shall pass the desktop benad 

}

class Palmtop extends Computer{
    // we define a constructor  through which we shall pass the desktop benad 
 
}

class Tablet extends Computer{
    // we define a constructor  through which we shall pass the desktop benad 

}



// acess methods



